src/libsyntax_ext/format.rs

```diff
@@ -131,24 +145,49 @@ fn parse_args(ecx: &mut ExtCtxt, sp: Span, tts: &[ast::TokenTree])
                     ecx.struct_span_err(e.span,
                                         &format!("duplicate argument named `{}`",
                                                  name))
-                       .span_note(prev.span, "previously here")
+                       .span_note(args[*prev].span, "previously here")
                        .emit();
                     continue
                 }
             }
-            order.push(name.to_string());
-            names.insert(name.to_string(), e);
+
+            // Resolve names into slots early.
+            // Since all the positional args are already seen at this point
+            // if the input is valid, we can simply append to the positional
+            // args. And remember the names.
+            let slot = args.len();
+            names.insert(name.to_string(), slot);
+            args.push(e);
         } else {
             args.push(panictry!(p.parse_expr()));
         }
     }
-    Some((fmtstr, args, order, names))
+    Some((fmtstr, args, names))
 }
 
 impl<'a, 'b> Context<'a, 'b> {
-    /// Verifies one piece of a parse string. All errors are not emitted as
-    /// fatal so we can continue giving errors about this and possibly other
-    /// format strings.
+    fn resolve_name_inplace(&self, p: &mut parse::Piece) {
+        let lookup = |s| *self.names.get(s).unwrap_or(&0);
```

Comment:
It's taken during parse of illegal invocations like in `format_args!("{foo}")` where the piece tries to reference a non-existent named argument, so this is necessary. Maybe I should add a comment regarding this.